pkg/backup/rbac.go

```diff
@@ -0,0 +1,194 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	arkdiscovery "github.com/heptio/ark/pkg/discovery"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	rbac "k8s.io/api/rbac/v1"
+	rbacbeta "k8s.io/api/rbac/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	rbacclient "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	rbacclientbeta "k8s.io/client-go/kubernetes/typed/rbac/v1beta1"
+)
+
+type CRBLister interface {
+	// List returns a slice of ClusterRoleBindings which can represent either v1 or v1beta1 ClusterRoleBindings.
+	List() ([]ClusterRoleBinding, error)
+}
+
+type V1CRBLister struct {
+	client rbacclient.ClusterRoleBindingInterface
+}
+
+func (v1cl V1CRBLister) List() ([]ClusterRoleBinding, error) {
+	var crbs []ClusterRoleBinding
+
+	crbList, err := v1cl.client.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	for _, crb := range crbList.Items {
+		crbs = append(crbs, v1crb{crb: &crb})
+	}
+
+	return crbs, nil
+}
+
+type V1beta1CRBLister struct {
+	client rbacclientbeta.ClusterRoleBindingInterface
+}
+
+func (v1b1 V1beta1CRBLister) List() ([]ClusterRoleBinding, error) {
+	var crbs []ClusterRoleBinding
+
+	crbList, err := v1b1.client.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	for _, crb := range crbList.Items {
+		crbs = append(crbs, v1beta1crb{crb: &crb})
+	}
+
+	return crbs, nil
+}
+
+func NewCRBLister(log logrus.FieldLogger, clientset kubernetes.Interface) (CRBLister, error) {
+	var lister CRBLister
+	discoveryHelper, err := arkdiscovery.NewHelper(clientset.Discovery(), log)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	var supportedAPI metav1.GroupVersionForDiscovery
+	apigroups := discoveryHelper.APIGroups()
+	for _, ag := range apigroups {
+		if ag.Name == "rbac.authorization.k8s.io" {
+			supportedAPI = ag.PreferredVersion
+			break
+		}
+	}
+
+	switch supportedAPI.Version {
+	case "v1":
+		lister = V1CRBLister{client: clientset.RbacV1().ClusterRoleBindings()}
+	case "v1beta1":
+		lister = V1beta1CRBLister{client: clientset.RbacV1beta1().ClusterRoleBindings()}
+	// Return an empty client if we don't have a known version.
+	default:
+	}
+	return lister, nil
+}
+
+// ClusterRoleBinding abstracts access to ClusterRoleBindings whether they're v1 or v1beta1.
+type ClusterRoleBinding interface {
+	// Name returns the name of a ClusterRoleBinding.
+	Name() string
+	// Subjects translates a ClusterRoleBindings Subjects slice into a slice of subject wrapper objects.
+	Subjects() []subject
```

Comment:
One thing you could do is instead of exposing `Subjects()` and needing the wrapper type for that, just add a method to the `ClusterRoleBinding` interface called `HasSubject` that takes a subject kind, namespace, and name, and returns true or false -- since that's all we really need. That would minimize the amount of this wrapper code that we need

Reply:
Actually may need to be `HasServiceAccountSubject` so it doesn't have to take the kind (since that's part of the API group/version), but same idea.